### PR TITLE
Feat/esm tests

### PR DESCRIPTION
**Pull request by mandresm**
ESM-Tests final implementation, including check compilations and runs (comparing files with the last test) for CI.

**TODO**

Strikeout items are to be solved after the merge as the merge is urgent.

- [x] Docstring all
- [x] Finish cleanup
- [ ] <del>Documentation
- [x] Implement jobs for all computers
  - [x] ollie
  - [x] levante
  - [x] hlrn
  - [x] aleph
  - [x] juwels
- [ ] <del>ESM-Tests status (`esm_tests -t`) pushed somewhere for visualization (home at github/readthedocs?)
- [x] ESM-Tests status triggered with `esm_tools --test-state`

### CODE REVIEW

**Comment by pgierz** [comment]
Hooray for auto-tests! I'll have some time to look over this after the lunch break

**Comment by mandresm** [comment]
It's a very complicated mess with a lot of tricks to not change the ESM-Tools source code but still be able to test it for consistency... but it works :D

File: src/esm_tests/cli.py
Comment (pgierz) [inline src/esm_tests/cli.py:67]: You can probably remove this; if it isn't caught in a variable it will just land in memory nirvana, or does the `user_info` function do something magical to the `info` dictionary?

File: src/esm_tests/cli.py
Comment (pgierz) [inline src/esm_tests/cli.py:87]: maybe with a `get`? I don't know the codebase enough, but just ideas for simple things that might otherwise explode

**Comment by mandresm** [comment]
In fact, the test fail because it is finding the change in the FESOM-REcoM branch that happen yesterday in release XD

**Comment by pgierz** [comment]
Do you want feedback on src for esm-tests? I can give some ideas at least on the changes in this PR; and I can clear off a day to go through the whole source code, but only if you want/feel that it is needed. If we are all happy saying "it works, don't touch it" that's fine as well :-)

**Comment by mandresm** [comment]
Up to you, feedback is always good and I can learn a lot, if you feel you have time for it go for it!

**Comment by denizural** [comment]
A quick question. Is there a specific reason to use `.yml` instead of `.yaml` extension? It looks like these are GitHub actions files, right?

**Comment by mandresm** [comment]
I think they can have both extensions (https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#about-yaml-syntax-for-workflows), and I think it was @pgierz that wrote the first workflow, so I just kept it the same. I think it's because the default workflow provided by GitHub is named with `.yml` so we all followed

**Comment by pgierz** [comment]
...yup I just followed the docu example.

File: src/esm_runscripts/namelists.py
Comment (denizural) [inline src/esm_runscripts/namelists.py]: Won't change the functionality but I would do it like
```python
fesom_iceberg_dir = config["fesom"].get("iceberg_dir", None)
```
since this is used multiple times after that. This would decrease the code repetition.

File: src/esm_tests/tests.py
Comment (denizural) [inline src/esm_tests/tests.py]: For a future code refactoring, these lines are a good candidate for using `enums`:
```python
from enum import Enum

class States(Enum):
    RUNS = 0
    RUN_FILES_IDENTICAL = 1
    RUN_FILES_DIFFER = 2
    SUBMISSION_FAILED = 3
    RUN_FILES_NOT_FOUND = 4
    RUN_FAILED = 5
``` 
You can then use these constants instead of a dictionary. It will also make it more compatible across other functions. But at this point it is not that important.

File: src/esm_runscripts/namelists.py
Comment (mandresm) [reply src/esm_runscripts/namelists.py]: Modified as suggested

File: src/esm_tests/tests.py
Comment (mandresm) [reply src/esm_tests/tests.py]: Can you clarify how would you use `enum` to improve this part of the code? I am not able to understand it just with the info you provided. I understand the class you built, but how would you use it and how is it superior to the dictionary?

Comment (pgierz) [reply src/esm_tests/tests.py]: Short answer: you save some square brackets and quotes, and it's more "readable", although that gets strongly into the personal opinion side of things. Downside is you introduce a construct not everyone is familiar with.

I am sure there are more reasons for enum being a nice thing (there is for instance an automatic way to create [automatic values](https://docs.python.org/3/library/enum.html#using-automatic-values) ), but unless you want to try out some new code karate it is at this level rather redundant.


- - - -

The above is an outdated comment from @denizural's outdated `enum` suggestion

Comment (pgierz) [reply src/esm_tests/tests.py]: @denizural if you have suggestions like this, why not directly add them to Miguel's branch?
- - - -
The above is also an outdated comment from @denizural enum suggestion.

**Comment by mandresm** [comment]
Hi everyone, something I have in mind about `esm_tests`. Appart from doing the tests (either checks and actual tests), storing all the configuration files and comparing against old ones, `esm_tests` also stores the information of the last actual tests in `esm_tools/src/resources/state.yaml`. This information can be printed with `esm_tests -t` (see image below). I think this information is very valuable for the user, as it shows what is supported and when was last tested successfully (or not). I have several questions for you about it:

1. Should we add an `esm_tools --state` command  (that basically links `esm_tests -t`)  so that the users can get that info from ESM-Tools without having to know anything about `esm_tests`?

2. Should we add this information anywhere in the documentation/github homepage? If the answer is yes, and you have some suggestions for a format (table/list/status images...) please let me konw.

Screen Shot 2022-06-13 at 11 01 45

**Comment by denizural** [comment]
This idea sounds good. It is called `regression` in software testing. 

Maybe we can host this information on GitHub and and when the command is invoked, `esm_tools` can grab the info from GitHub (read only for the users, only developers can write) and display on the screen.

I would call it something like `--test-state`, btw just to reserve more generic `state` for some future work. But of course, it is up to you to choose a name.

File: src/esm_runscripts/namelists.py
Comment (pgierz) [inline src/esm_runscripts/namelists.py]: I am not sure I like this, since the `check` mode might be on also for other reasons. Can we make it so that `esm-tests` gets its own flag for `esm-runscripts`?

File: src/esm_runscripts/namelists.py
Comment (pgierz) [inline src/esm_runscripts/namelists.py]: Sorry but...em....what? I trust you that this works, but what wizardry is this? You are counting the number of non-empty path elements, but then the last one, and then the last one after splitting the `_`? 

Even if it never breaks, someone at some point will stumble across this code and be highly confused. Just put a comment or something that shows what number this gets. I assume it is something like:
```
/some///./crazy/path/with///a/new/file_38125
```
which yields
```
38125
```

File: src/esm_tests/cli.py
Comment (pgierz) [inline src/esm_tests/cli.py:74]: Thank you!! We should always opt rather for erasing code than commenting it out, in my view. Code can be retrieved from the git history if you really need.

File: src/esm_tests/cli.py
Comment (pgierz) [inline src/esm_tests/cli.py]: OK, this already answers my comment from further above.

File: src/esm_tests/cli.py
Comment (pgierz) [inline src/esm_tests/cli.py]: As above, this works, but looks confusing. Probably it would be nice to have something in the `helpers.py` which can mutate paths. I have no problem there including computer-specific functions. As an example:
```python
def clean_computer_specific_paths(path, computers=None):
    if computers is None:
        # Or some sensible default:
        computers = ["mistral", "levante"]:
    # ....something that grabs the computer config yaml .....
    for computer_config in computer_configs:
        for remove_prefix in computer_config["allowed_removes"]:
            path = path.replace(remove_prefix, "")
    return path



```

File: src/esm_tests/cli.py
Comment (pgierz) [inline src/esm_tests/cli.py]: Could be included as `.get("account")` and then this defaults to `None`

File: src/esm_tests/repos.py
Comment (pgierz) [inline src/esm_tests/repos.py:17]: It may be nice to include some example of which info must be included. That can go either (or both) in the function body at the beginning of this function which throws "friendly" key errors, and/or in the doc-string.

File: src/esm_tests/repos.py
Comment (pgierz) [inline src/esm_tests/repos.py:59]: I am unsure how difficult this is, but it might be interesting to create "custom" Sphinx headings. @denizural is correct, this heading should not be `Raises` as we do not raise an error here, but it is an `Error` that might be usefully separated out in the documentation to show what sort of things break this function.

Personally I would just raise an exception, but users don't like long scary tracebacks...

**Review by pgierz — COMMENTED** [review]
Some small comments and ideas, up to you to include them or not. To be honest, I would say: put this in, and we can separately include the clean-up, changes, and/or my ideas; as it is a valuable feature to have in CI as soon as possible to avoid future headaches.

Excellent work with the tests Miguel 👍🏻

File: src/esm_runscripts/namelists.py
Comment (pgierz) [reply src/esm_runscripts/namelists.py:498]: The above is also an outdated comment from @denizural `enum` suggestion.

**Comment by pgierz** [comment]
A brief note: I am not sure why exactly, but it put comments of some ideas I had to @denizural notes in their twice. I marked those answers in my own review as redundant. Sorry for the confusion. GitHub confuses me sometimes, maybe I am slowly getting old...

File: src/esm_runscripts/namelists.py
Comment (denizural) [reply src/esm_runscripts/namelists.py]: Rather than using string functions we should use Pathlib or os.path methods. Eg. `parent`, ...

File: src/esm_tests/repos.py
Comment (denizural) [reply src/esm_tests/repos.py:59]: We can always catch the exception with `user_error` to suppress the traceback & exit gracefully.

File: src/esm_tests/cli.py
Comment (mandresm) [reply src/esm_tests/cli.py]: `ACCOUNT` should be removed here as it's not in used and shouldn't be used anymore.

File: src/esm_tests/repos.py
Comment (mandresm) [reply src/esm_tests/repos.py:17]: `info` is a dictionary cotainning most of the information of the tests. It is build sequentially through the `cli.py`, starting with the loading of the `argpars` variables, and continuing with all the functions of esm_tests. It's similar to the `config` dictionary/object in ESM-Tools. Therefore, the description should be in a more general place, like in the `cli.py` right? Or maybe I just need to be build an `dict` subclass and put the docstrings there?

File: src/esm_runscripts/namelists.py
Comment (mandresm) [reply src/esm_runscripts/namelists.py]: Yeah, this is horrible... I'll find a better way to fix this.

File: src/esm_runscripts/namelists.py
Comment (mandresm) [reply src/esm_runscripts/namelists.py]: The minimum would have been to use regex, but I don't know what I was doing that day. In any case, I don't like this either. I will probably revert most of the `namelists.py` changes and fix it from the resources side (avoiding being so so invasive to the ESM-Tools code)

File: src/esm_tests/repos.py
Comment (pgierz) [reply src/esm_tests/repos.py:17]: A dict subclass is a nice idea here. You can have it fail during initialisation if it does not have the proper parts.

File: src/esm_runscripts/namelists.py
Comment (pgierz) [reply src/esm_runscripts/namelists.py]: Assuming the number is always the last thing in the file name:
```suggestion
                    ib_num_new = int(iceberg_dir.split("_")[-1])
```

File: src/esm_tests/cli.py
Comment (mandresm) [reply src/esm_tests/cli.py]: The user and computer specific parts have been generalized and included into a method.

File: src/esm_runscripts/namelists.py
Comment (mandresm) [reply src/esm_runscripts/namelists.py]: solved by reverting the changes in `namelist.py` and handling all throught esm_tests

**Comment by mandresm** [comment]
I think all points have been tackled

File: src/esm_tests/info.py
Comment (pgierz) [inline src/esm_tests/info.py:66]:
```suggestion
        -------
         Returns a tuple with the following entries:
```

Check how that renders in the documentation.

**Review by pgierz — CHANGES_REQUESTED** [review]

File: src/esm_tests/info.py
Comment (mandresm) [reply src/esm_tests/info.py:66]: The dosctring building in the docs is broken since long time, so I cannot check it. I tried a couple of times to try to fix it and failed.

**Comment by mandresm** [comment]
After passing the black formatting I'll be ready to merge. I'll work on the documentation later, but at least every method/class has now detailed docstrings. I'll need an approval before merging ;)

**Comment by mandresm** [comment]
#bump

**Comment by mandresm** [comment]
#bump

**Pull request merged**
